File: codes/lua/004_loops.lua
function sprint(...)
    local rest_arguments = {...}
    local new_array = {}
    for _, argument in ipairs(rest_arguments) do
        table.insert(new_array, tostring(argument))
    end
    print(table.concat(new_array, ""))
end

-- while loop

i1 = 1
while (true) do
    if (i1 == 2) then
        if (i1 >= 5) then break end
        i1 = i1 + 1
        goto next
    end
    sprint("while loop v1, i1 is ", i1)
    if (i1 >= 5) then break end
    i1 = i1 + 1
    ::next::
end

i2 = 1
while (i2 <= 5) do
    if (i2 == 2) then
        i2 = i2 + 1
        goto next
    end
    sprint("while loop v2, i2 is ", i2)
    i2 = i2 + 1
    ::next::
end

-- for loop

for i3 = 1, 10, 1 do -- start, stop, step
    if (i3 > 5) then break end
    if (i3 == 2) then goto next end
    sprint("for loop ascending, i3 is ", i3)
    ::next::
end

for i4 = 10, 1, -1 do -- start, stop, step
    if (i4 <= 5) then break end
    if (i4 == 9) then goto next end
    sprint("for loop descending, i4 is ", i4)
    ::next::
end

File: codes/r/004_loops.r
# while loop

i <- 1
while (TRUE) {
    if (i == 2) {
        if (i >= 5) break
        i <- i + 1
        next
    }
    cat(paste(sep = "", "while loop v1, i is ", i, "\n"))
    if (i >= 5) break
    i <- i + 1
}

i <- 1
while (i <= 5) {
    if (i == 2) {
        i <- i + 1
        next
    }
    cat(paste(sep = "", "while loop v2, i is ", i, "\n"))
    i <- i + 1
}

# for loop

for (i in (1:10)) { # (start, stop)
    if (i > 5) break
    if (i == 2) next
    cat(paste(sep = "", "for loop v1 ascending, i is ", i, "\n"))
}

for (i in (10:1)) { # (start, stop)
    if (i <= 5) break
    if (i == 9) next
    cat(paste(sep = "", "for loop v1 descending, i is ", i, "\n"))
}

for (i in seq(1, 10, by = 1)) { # (start, stop, by = step)
    if (i > 5) break
    if (i == 2) next
    cat(paste(sep = "", "for loop v2 ascending, i is ", i, "\n"))
}

for (i in seq(10, 1, by = -1)) { # (start, stop, by = step)
    if (i <= 5) break
    if (i == 9) next
    cat(paste(sep = "", "for loop v2 descending, i is ", i, "\n"))
}
